### PR TITLE
Enable mTLS for downstream clients

### PR DESCRIPTION
**Pull request by alanrobinson-dwp**
### Type of change

- Enhancement / new feature

### Description

_This PR is to introduce functionality so that configuration can be set to make downstream clients require authentication when connecting to Kroxylicious_

### Additional Context

_Current functionality will only set up one way TLS, and from a security standpoint using Kroxylicious in our environments we require mTLS. The downstream SSL builder only loads the key details from the clusters tls configuration and ignores the trust. I've made changes to load the trust if provided and set the client authentication to required in this instance. I make the assumption that if someone is providing the trust configuration for downstream then they must require mTLS_

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [ ] Update documentation
- [x] Make sure all unit/integration tests pass
- [x] Make sure all Sonarcloud warnings are addressed or are justifiably ignored.
- [ ] If applicable to the change, [trigger the system test suite](../blob/main/DEV_GUIDE.md#jenkins-pipeline-for-system-tests).  Make sure tests pass.
- [ ] If applicable to the change, [trigger the performance test suite](../blob/main/PERFORMANCE.md#jenkins-pipeline-for-performance). Ensure that any degradations to performance numbers are understood and justified.
- [x] Ensure the PR references relevant issue(s) so they are closed on merging.
- [x] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

> **_NOTE:_**  You must be a member of `@kroxylicious/developers` to trigger the system test and performance test suites.  If you are not part of this group, comment on the PR requesting a trigger, tagging `@kroxylicious/developers`.

### CODE REVIEW

**Comment by ddaviesbrackett-dwp** [comment]
This is to address issue https://github.com/kroxylicious/kroxylicious/issues/1561, and also may influence https://github.com/kroxylicious/kroxylicious/issues/1577.

**Comment by k-wall** [comment]
@alanrobinson-dwp thank you for opening the PR.

>  I make the assumption that if someone is providing the trust configuration for downstream then they must require mTLS

I wonder about this.  Kafka itself defaults to `none` and lets the user [configure](https://kafka.apache.org/documentation/#brokerconfigs_ssl.client.auth) `required` or `requested`.  I think it would be least surprise if Kroxylicious adopted the same behaviour and used the same defaulting.  This would require a new piece of TLS config (probably `clientAuth`).  OTOH you could argue from a secure by default standpoint, `requested` ought to be default, with configuration relaxing it down.

Separately, I think the change should be supported by integration test(s) - probably new tests in `io.kroxylicious.proxy.TlsIT` would be appropriate.

**Comment by alanrobinson-dwp** [comment]
Hi Keith, thanks for comments and taking a look. 

We'd had the same discussions with regards to how the brokers do it and possibly having an extra variable of clientAuth in either of Tls or Truststore record types then mapping up "required" or "requested" with the ClientAuth ENUM - NONE, OPTIONAL, REQUIRE; values but we erred on the lighter touch.

Do you have a preference with regards to your security standpoint? From our viewpoint, secure by default fits our requirements better, but we're happy to go with what suits with Kroxy better.

For the setting, we could add a clientAuth into Tls.java? Perhaps a boolean requiresAuth that will return REQUIRE if trust is != null and clientAuth is not set?

**Comment by SamBarker** [comment]
Thanks for the pr @alanrobinson-dwp! 

I think secure by default is a good posture, but I'm not entirely sure it a stance a proxy can take unilaterally as one of our goals is to be transparent to clients & brokers (which I take to mean works by default). 

I don't know enough about `m_tls` and its general usage to have a strong view. I do however think defaulting to `REQUIRED` when a `TrustStore` is present along with an escape hatch to control the clientTrust property when needed.  

>For the setting, we could add a clientAuth into Tls.java? Perhaps a boolean requiresAuth that will return REQUIRE if trust is != null and clientAuth is not set?

I'd err towards just exposing the clientTrust enum rather than trying to infer what todo based on other fields. I think thats easier for users to reason about and is closer to how Kafka does it (be that right or wrong) and easier to validate for example user said REQUIRED but didn't provide a trust store - cluster doesn't load. However I'm happy to take direction from others with more experience in dealing with TLS.

**Comment by alanrobinson-dwp** [comment]
Hi @SamBarker @k-wall 

We had lots of philosophical conversations yesterday in the office about best way to implement this. I agree it'd be tidier to map config directly to the ClientAuth Enum. I do wonder what Kafka does behind the scenes.

We landed though on following exactly as broker does it (https://kafka.apache.org/090/documentation.html) with "none" being default. Our thinking being that if a team were used to setting "required" or "requested" when speaking with brokers the client properties wouldn't change much when switching to Kroxylicious. 

The new way has been added to PR if you wouldn't mind giving some feedback? I'll squash and tidy commits before going any further.

I'm currently getting my head around the TlsIT.java class to look into covering this with an IT test. I've not used these Fluent config builder classes before so just familiarising with that before writing test.

Thanks
Alan

**Comment by SamBarker** [comment]
Thanks @alanrobinson-dwp I'll review in my morning! 

Building a proxy is a remarkably philosophical exercise. I'm glad to hear that it sounds like you've landed in a similar place to my thinking.

File: kroxylicious-api/pom.xml
Comment (k-wall) [inline kroxylicious-api/pom.xml:75]: The aim is to keep the use of the netty an implementational concern known only to the internals of the Proxy.  So we don't want to bleed Netty API into the Filter API.     So my suggestion is that we have our own ClientAuth enum and use that in the API.  Let the implementation map from that to the Netty one.

Comment (alanrobinson-dwp) [reply kroxylicious-api/pom.xml:75]: Thanks Keith, makes sense, will sort that out

Comment (alanrobinson-dwp) [reply kroxylicious-api/pom.xml]: Hi @k-wall I've done an update there to have a mapper using an Enum in the api that maps between and removes that dependency. Is that closer to what you had in mind?

File: kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/model/VirtualClusterTest.java
Comment (SamBarker) [inline kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/model/VirtualClusterTest.java:45]: I appreciate you are following the existing pattern here but I've been preferring `satisfies` over extracting recently. So I started trying to write a fuller test locally using that. 
```suggestion
        assertThat(virtualCluster)
                .isNotNull()
                .satisfies(vc ->
                        assertThat(vc.getDownstreamSslContext())
                                .isPresent()
                                .hasValueSatisfying(
                                        sslContext -> {
                                            assertThat(sslContext.isClient()).isFalse();
                                            assertThat(sslContext.isServer()).isTrue();
                                            final SSLEngine sslEngine = sslContext.newEngine(ByteBufAllocator.DEFAULT);
                                            assertThat(sslEngine.getWantClientAuth()).isTrue(); //Depending on ClientAuth config
                                            assertThat(sslEngine.getNeedClientAuth()).isTrue(); //Depending on ClientAuth config
                                        }))
                .extracting("upstreamSslContext").isNotNull();

```

Which lead me to a couple of realisations. 

1. the `isNotNull` assertion isn't very helpful we should be checking the Optional isPresent instead (thats an existing bug)
2. This stuff is hard to test should but given the variability of `want` or `need` client auth I think it deserves its own test.

**Review by SamBarker — COMMENTED** [review]
Thanks @alanrobinson-dwp I think the test coverage needs a bit more work.

I'm not entirely convinced by `TlsClientAuth` but I need to mull that one over a bit further. I suspect we should be doing something in the runtime to map from our enum to the netty names rather than implicitly coupling our API to theirs via strings.

File: kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/config/tls/NettyTrustProviderTest.java
Comment (alanrobinson-dwp) [reply kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/config/tls/NettyTrustProviderTest.java:123]: yeah that's a fair comment, I had that thought when putting them in. Can strip out

**Comment by SamBarker** [comment]
build is failing with
```
Error:  Failed to execute goal com.mycila:license-maven-plugin:4.6:check (license-check) on project kroxylicious-api: Some files do not have the expected license header. Run license:format to update them. -> [Help 1]
```

Assuming your branch is up to date with `main` running the full maven build locally will resolve that.

File: kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/model/VirtualClusterTest.java
Comment (alanrobinson-dwp) [reply kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/model/VirtualClusterTest.java:45]: Yeah, I'd tried to keep it light touch originally as new to the code, but can take another look at the tests.

Main issue I had yesterday was expecting the build to generate `withClientAuth` and `withNewClientAuth` methods on the config builders off the back of `withNewTls` builder method based on the new `clientAuth` String I'd added into Tls.java so I could do the integration tests, but it wasn't being generated. I was trying to work out if I'd missed something there, but I'll wait until you have a think on how to map our configuration with the `ClientAuth` netty Enum

**Comment by SamBarker** [comment]
I've opened https://github.com/alanrobinson-dwp/kroxylicious/pull/1/files 😬  

It started off fleshing out the tests in https://github.com/alanrobinson-dwp/kroxylicious/pull/1/commits/a39cbf15e61613951c2daa6fa915af3d20b64b76 which worked on the code here, and meant it was safe to go refactoring.

As @k-wall mentioned we don't want to leek the Netty API into our API, and while your proposal stopped the explicit linkage it didn't actually break the connection. It just coupled us by value. So I changed that by performing the mapping in code that actually knows about Netty. 

With that in place I went a bit further and said that we should validate that the configuration makes sense. There is no point in users configuring `clientAuth: required` and then not supplying a trustStore to validate certificates. 

I think the changes ended up being more invasive than you were trying for and certainly more invasive than I intended to be to your work but I think this is what we need. Would love to get everyones thoughts on that PR (merge into this one if your happy with them).

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java
Comment (k-wall) [inline kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java:65]: At first glance, I assumed the return value was whether client auth was being enforced or not.    It is not, it whether the object is in a validate state.  Which is always true, otherwise the object would fail construction.

I'm not really appreciating the value of exposing this method for test.  I'd just inline it.

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/TlsClientAuth.java
Comment (k-wall) [inline kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/TlsClientAuth.java:17]: I think these methods and fields (clientAuth and nettyClientAuth) of the enum are now dead.

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/TlsClientAuth.java
Comment (k-wall) [inline kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/TlsClientAuth.java:1]: Missing licence

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/VirtualCluster.java
Comment (k-wall) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/VirtualCluster.java:199]: I like the fact that it is this module that knows how to map our ClientAuth values into the Netty ones.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/VirtualCluster.java
Comment (k-wall) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/VirtualCluster.java:211]: Java 17's yield exhaustiveness means we can turn the runtime error into a compile time one.

```suggestion
    private static ClientAuth toNettyClientAuth(Tls tlsConfiguration) {
        return switch (tlsConfiguration.clientAuth()) {
            case REQUIRED -> ClientAuth.REQUIRE;
            case REQUESTED -> ClientAuth.OPTIONAL;
            case NONE -> ClientAuth.NONE;
        };
    }
```

**Comment by k-wall** [comment]
I notice that there is no existing documentation covering how to configure trust on the downstream side.  Let's fix that on a separate PR.  It can document the new configuration introduced by this PR too.  Happy to see this PR merged without a doc update.

**Comment by alanrobinson-dwp** [comment]
Hi @k-wall, thanks for all your comments. I've done a couple of the quicker fixes (Tls redundant classes and simplified the switch). I'll let @SamBarker take a look at the comments around his additions.

I've noticed the commit history is a little messy now, how does this work on the PRs, are they all squashed on merge or do you want me to tidy them up once we're finished before any merge takes place?

I'm still a little stuck on the TlsIT tests and can't see the `withClientAuth` or `withNewClientAuth` methods I was expecting to be generated on the Tls record class. Any ideas why this might not be happening?

I've tested the current PR code set on our environments too and looking good so far.

Errors with `Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `io.kroxylicious.proxy.config.tls.TlsClientAuth` from String "required": not one of the values accepted for Enum class: [NONE, REQUESTED, REQUIRED]` when set and not an expected value.

Also my client can't connect when REQUIRED if it hasn't set it's ssl.keystore.location/password etc

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java
Comment (k-wall) [inline kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java:67]: Is this condition actually correct?  Why insist that the server overrides his trust anchors to use this feature?  What if he wants to use platform trust?

Using TlsClientAuth.REQUIRED or REQUESTED instructs the TLS server to include a Certificate Request in the TLS handshake.  The client then responds with something that proves he is in possession of the his key (using a signature) and sends his cert.  The server will then verify signature and the check cert against the trust anchor(s).

**Comment by k-wall** [comment]
> Hi @k-wall, thanks for all your comments. I've done a couple of the quicker fixes (Tls redundant classes and simplified the switch). I'll let @SamBarker take a look at the comments around his additions.
> 
> I've noticed the commit history is a little messy now, how does this work on the PRs, are they all squashed on merge or do you want me to tidy them up once we're finished before any merge takes place?

We haven't really a got fixed policy.  

When I am working on a PR I tend to squash together commits that fix up trivial stuff, but leave commits separate  once I start working answering review comments (to keep it easy for reviewers to see how I've addressed their comments).

When it comes to merging a PR back to main, I tend to squash down to single commit unless there are significant commits from separate authors.

I think your PR history is okay as it is.


> 
> I'm still a little stuck on the TlsIT tests and can't see the `withClientAuth` or `withNewClientAuth` methods I was expecting to be generated on the Tls record class. Any ideas why this might not be happening?
> 

Did you see https://github.com/alanrobinson-dwp/kroxylicious/pull/2


> I've tested the current PR code set on our environments too and looking good so far.
> 
> Errors with `Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type io.kroxylicious.proxy.config.tls.TlsClientAuth from String "required": not one of the values accepted for Enum class: [NONE, REQUESTED, REQUIRED]` when set and not an expected value.
> 
> Also my client can't connect when REQUIRED if it hasn't set it's ssl.keystore.location/password etc

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java
Comment (alanrobinson-dwp) [reply kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java:67]: This probably goes back to when I'd made the assumption that providing the trust meant it needed ClientAuth and there was a check on whether trust was provided before setting ClientAuth. Now we're setting trust regardless (even just to PlatformTrustProvider as default) it probably negates this check. 

Your public-CA scenario sounds like a use case for PlatformTrustProvider and ClientAuth REQUIRED.

Comment (k-wall) [reply kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java:67]: I'd like to hear @SamBarker thoughts.. It has been a long week and my brain is fried!

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java
Comment (SamBarker) [reply kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java:61]: Tom and I were chatting about defaults for a bit as well. 

The current round trip test is enforcing a stronger guarantee than we need. I was attempting to do the minimum in this pr. rather than bolt on a refactoring there.

Initially I defaulted null to none in the constructor but that broke configuration tests as the none was serialised back out which means it appears in the upstream tls nodes which is meaningless at best.

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java
Comment (SamBarker) [reply kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java:65]: Initially I had the validation independent of clothe ctor and thus had tests around it. When I moved it into the ctor I took the lazy option of keeping the method rather than rethinking how to assert the behaviour.

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/TlsClientAuth.java
Comment (SamBarker) [reply kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/TlsClientAuth.java:7]: Hmmmm that shouldn't be an issue the maven build should be adding them automatically...

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/VirtualCluster.java
Comment (SamBarker) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/VirtualCluster.java:199]: Yeah that's the core thing I was aiming for but couldn't think how to explain in comments without just writing the code to do it first 😆

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java
Comment (SamBarker) [reply kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java:67]: Arguably the check is moot/in appropriate/difficult to satisfy. I hesitated but I was persuaded by @tombentley that when it comes to security its better to err towards manual configuration. Rather than users getting certs validated by the platform trust store when they intend them to be validated by the custom trust store .e.g the proxy is misconfigured but still works by chance.

I do wonder how hard it would be in practice to specify the correct path for the platform trust store.

**Comment by SamBarker** [comment]
> Errors with `Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type io.kroxylicious.proxy.config.tls.TlsClientAuth from String "required": not one of the values accepted for Enum class: [NONE, REQUESTED, REQUIRED]` when set and not an expected value.

We need to enable `MapperFeature.ACCEPT_CASE_INSENSITIVE_ENUMS` to fix that, personally I think thats a good idea.

>Also my client can't connect when REQUIRED if it hasn't set it's ssl.keystore.location/password etc

Thats we are trying to achieve isn't it? Or are do you use platform trust for validating client certs?

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java
Comment (SamBarker) [inline kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java:43]: Hmmm. Thinking about this. Should we really deprecate and remove this signature given its used for upstream connections where client auth isn't currently supported?

**Review by SamBarker — APPROVED** [review]
Thanks @alanrobinson-dwp this has turned into a much longer process than any of expected.

I think is the last thing from me https://github.com/alanrobinson-dwp/kroxylicious/pull/4.

I'm no longer convinced about the depreciation of the `null` clientAuth constructor for Tls but happy to leave the final call on that to @k-wall. 

Approving now so I don't hold things up while traveling

**Comment by SamBarker** [comment]
>I've noticed the commit history is a little messy now, how does this work on the PRs, are they all squashed on merge or do you want me to tidy them up once we're finished before any merge takes place?

I don't generally worry about the git history as an artefact in its own right. I can't say I've ever gone through a code base commit by commit. I have stepped through the history of a class or block in IntelliJ to work out what the motivation for a change was, but even then I'm generally focused on the delta rather than the history itself. 

Personally I think squash merge looses a lot of detail (that is arguably noise) so I tend to just merge as is but I'm very aware I'm in a minority around this stuff. 😉

**Comment by alanrobinson-dwp** [comment]
> Thats we are trying to achieve isn't it? Or are do you use platform trust for validating client certs?

Exactly yeah, so working fine for us now with our requirements. We'd had a pretty rigorous IT Healthcheck on the lead up to Go Live and this was flagged after we were testing connections with different certificates using different CAs between our Client and Proxy. This ticks a big box for us!

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java
Comment (k-wall) [reply kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java:41]: Agreed, that makes sense to me.

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java
Comment (k-wall) [reply kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java:64]: > I hesitated but I was persuaded by @tombentley that when it comes to security its better to err towards manual configuration.

I can see the logic in that idea.

However, if we go down that route, doesn't that mean that `clientAuth` ought to be a property of the `TrustProvider` rather than `Tls`?    That way the configuration model enforces the business rule.

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java
Comment (k-wall) [inline kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java:20]:
```suggestion
 * @param clientAuth when in the node is in the TLS <em>server</em> role, clientAuth specifies how the server should authenticate clients. This option has no meaning in the node is in the TLS <em>client</em> role. 
```

We've used the expression TLS server and TLS client elsewhere in the Javadoc to describe how options apply to the different roles.  I think it make sense to adopt the same language here.

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/TlsClientAuth.java
Comment (k-wall) [inline kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/TlsClientAuth.java:15]: Fields clientAuth and nettyClientAuth are unused.  We should remove them.

File: kroxylicious-integration-test-support/src/test/java/io/kroxylicious/proxy/config/ConfigurationTest.java
Comment (k-wall) [inline kroxylicious-integration-test-support/src/test/java/io/kroxylicious/proxy/config/ConfigurationTest.java:134]: These are upstream test cases.  The TLS peer is the client role so the clientAuth setting has no meaning.  It seems weird/confusing to see this configuration.

File: kroxylicious-integration-test-support/src/main/java/io/kroxylicious/proxy/config/BuilderConfig.java
Comment (k-wall) [inline kroxylicious-integration-test-support/src/main/java/io/kroxylicious/proxy/config/BuilderConfig.java:24]:
```suggestion
```

@SamBarker is right, this line is superfluous.  We should take it away.

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java
Comment (k-wall) [reply kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java:64]: @alanrobinson-dwp @SamBarker Following that idea,  I tried pushing clientAuth to the TrustProvider. WDYT?

https://github.com/alanrobinson-dwp/kroxylicious/pull/5

Comment (alanrobinson-dwp) [reply kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java:64]: I'd been flip flopping for a bit on where to put the clientAuth value and we had a few discussions here. Originally TrustProvider seemed most obvious being that it is only needed if set. We ended up following the broker though which has settings

ssl.client.auth
ssl.keystore.*
ssl.truststore.*

properties, so assuming Tls.java was to replicate the ssl level we landed there. It makes just as much sense in TrustProvider mind.

I had for a split second considered the benefits of having separate Upstream and Downstream TLS classes, but then quickly backed away haha

File: kroxylicious-integration-test-support/src/test/java/io/kroxylicious/proxy/config/ConfigurationTest.java
Comment (SamBarker) [reply kroxylicious-integration-test-support/src/test/java/io/kroxylicious/proxy/config/ConfigurationTest.java:134]: Oh rubbish your right I'd missed they were upstream config.

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java
Comment (SamBarker) [reply kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java:64]: >I'd been flip flopping for a bit on where to put the clientAuth value and we had a few discussions here. Originally TrustProvider seemed most obvious being that it is only needed if set. We ended up following the broker though which has settings

Good to hear its not just me thats having trouble with this, and I really appreciate you taking the time to think it through rather than just pushing up the first thing you came up with.

>properties, so assuming Tls.java was to replicate the ssl level we landed there.

I certainly agree with starting from how does Kafka itself model this as thats what a lot of users will expect but I'm often left a bit confused and frustrated by how Kafka models a lot of config so its not the be all and end all.

>I had for a split second considered the benefits of having separate Upstream and Downstream TLS classes, but then quickly backed away haha

Yeah, I can see why you did but it does echo my thinking from https://github.com/alanrobinson-dwp/kroxylicious/pull/5#discussion_r1868671250 the appeal of Upstream and Downstream is that they are what we want to talk about so in Domain Driven Design terms they are part of our ubiquitous language. Which makes me think its the right direction to be moving. I don't have strong opinions if that happens in this PR or not but we if we want to get to `*stream` types sooner is better so that we have the right shape in the API we release.

**Comment by k-wall** [comment]
@alanrobinson-dwp hopefully we are almost there on this PR.
* Could you add an entry to the CHANGELOG.md
* Add yourself to the CONTRIBUTORS.md file if you wish.

I'll do one more review once that is done and hopefully merge it straight after.

**Comment by alanrobinson-dwp** [comment]
Thanks @k-wall I've added on change details and rebased. I'd feel a bit of a fraud putting my name on the contributors at the moment haha!

There is another ticket on our board for getting involved with a ciphers change, so maybes after that!

**Review by k-wall — APPROVED** [review]

**Comment by k-wall** [comment]
thanks for the contribution @alanrobinson-dwp

**Comment by SamBarker** [comment]
>Thanks @k-wall I've added on change details and rebased. I'd feel a bit of a fraud putting my name on the contributors at the moment haha!

I appreciate the modesty @alanrobinson-dwp but the change is yours as far as GitHub and I are concerned. I'm very glad you got this organised.

**Comment by k-wall** [comment]
> > Thanks @k-wall I've added on change details and rebased. I'd feel a bit of a fraud putting my name on the contributors at the moment haha!
> 
> I appreciate the modesty @alanrobinson-dwp but the change is yours as far as GitHub and I are concerned. I'm very glad you got this organised.

Absolutely, I feel exactly the same way.

**Pull request merged**
